collections/_posts/2022-10-03-MomentumResidualNeuralNetwork.md

```diff
@@ -0,0 +1,151 @@
+In classical ResNets, we have the feedforward relation:
+
+$$x_{n+1}=x_n+ f(x_n,\theta_n)$$
+
+where $$\theta_n$$ is the set of parameters, and $$x_n$$ the activations at layer $$n$$. 
+
+Memory issues occur when increasing the number of layers.
+
+The authors propose to use momentum equations that replace the classical relation above:
+
+$$ v_{n+1}=\gamma v_n+(1-\gamma)f(x_n,\theta_n)$$
+
+$$x_{n+1}=x_n+v_{n+1}$$
+
+where $$v$$ is a velocity term and $$\gamma$$ a momentum term.
+
+The method consists in modifying the forward equations using the same parameters as inputs. This is invertible since we can recover the values of $$x_{n}$$ and $$v_{n}$$ using $$x_{n+1}$$ and $$v_{n+1}$$.
+If we invert these equations, we get :
+
+$$x_n=x_{n+1}-v_{n+1}$$
+
+$$v_n=\frac{1}{\gamma} (v_{n+1}-(1-\gamma)f(x_n,\theta_n))$$
+
+This avoids the memory issues occuring due to the backpropagation step.
+
+Note: Momentum gradient descent is an alternative to classical gradient descent algorithms using second order partial derivatives.
+A overview of gradient descent optimization algorithms is given in the paper of S.Ruder (http://arxiv.org/abs/1609.04747) as mentioned in the paper.
+
+## Memory cost
+For usual ResNets, one need to store the weights of the network and the values of all activations for the training set at each layer. The memory needed is $$O(k*d*n_{batch})$$ while for the Momentum ResNets the memory need is $$O((1-\gamma)*k*d*n_{batch})$$.
+
+![](/collections/images/MomentumResNets/fig_memory_momentum.jpg)
+
+## The role of momentum
+When $$\gamma=0$$, they get a Classical ResNet. When $$\gamma \rightarrow 1$$, they get a special case of the invertible RevNet (Gomez et al, 2017). 
+
+The advantage of the Momentum ResNet compared to RevNet where two learnable functions are used, is its stability for convergence (proofs given in the paper).
+
+## Link with ODE (Ordinary Differential Equations)
+
+![](/collections/images/MomentumResNets/fig_momentum_odes.jpg)
+ 
+ResNets can be interpreted as a first order ODE. Indeed the term $$x_{n+1} - x_n$$ can be seen as the discretized partial derivative $$\dot{x}$$.
+
+$$ \dot{x} \quad  \rightarrow  \quad x_{n+1}-x_{n}$$
+
+Momentum ResNets can be interpreted as second-order ODEs by taking $$\epsilon=\frac{1}{1-\gamma}$$, they get:
+
+$$ v_{n+1}=v_{n} + \frac{f(x_n,\theta_n) - v_n}{\epsilon} $$
+
+$$ x_{n+1}=x_n+v_{n+1}$$
+
+
+by replacing the term $$v_n$$, they find:
+
+$$x_{n+1}-x_{n}=x_n-x_{n-1} + \frac{f(x_n,\theta_n)-v_n}{\epsilon}$$
+
+The second order derivative $$\ddot{x}$$ can be discretized using $$x_{n+1}-2x_n+x_{n-1}$$ and then we can interpret the Momentum ResNets as a second order ODE of the form:
+
+$$\epsilon \ddot{x}+\dot{x}=f(x,\theta)$$
+
+Briefly : "In the same way that ResNets can be seen as discretization of first order ODEs, Momentum ResNets can be seen as discretization of second-order ones."
```

Comment:
I think a neat formatting trick to better emphasize this sentence would be to format at it has a quote (using ">" at the begining of the line):
```suggestion
>In the same way that ResNets can be seen as discretization of first order ODEs, Momentum ResNets can be seen as discretization of second-order ones.
```